# Stats701 Homework 10, Winter 2018
### David Guo
#### davidguo@umich.edu

Discussion: Problem 1, discussed with Jeanhee Pak. Problem 2, discussed with Jeanhee Pak, Noah Gale. Problem 3, discussed with Noah Gale. Problem 4, discussed with Jeanhee Pak and Noah Gale.

Time to do each homework problem:

- Problem 1 took about 15 minutes

- Problem 2 took about 15 hours
    
- Problem 3 took about 5 hours

- Problem 4 took about 20 hours

- Write-up took about 3 hours

### 1

#### 1.1

In [1]:
import numpy as np

In [60]:
# top to bottom of height, towards t shadow, towards f shadow
np.array([[[0,0,1],[0,0,1],[0,0,1],[1,1,1]], #top slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]], #2nd slice
          [[0,0,0],[0,0,0],[0,1,1],[0,0,0]], #middle slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]], #4th slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]]]) #bottom slice

array([[[0, 0, 1],
        [0, 0, 1],
        [0, 0, 1],
        [1, 1, 1]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 1],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 1, 1],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 1],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 1],
        [0, 0, 0]]])

In [61]:
import tensorflow as tf

In [62]:
logo = tf.constant([[[0,0,1],[0,0,1],[0,0,1],[1,1,1]], #top slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]], #2nd slice
          [[0,0,0],[0,0,0],[0,1,1],[0,0,0]], #middle slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]], #4th slice
          [[0,0,0],[0,0,0],[0,0,1],[0,0,0]]], dtype = tf.float32)

In [5]:
logo

<tf.Tensor 'Const:0' shape=(5, 4, 3) dtype=float32>

In [6]:
sess = tf.Session()

In [7]:
with sess.as_default():
    print(logo.eval())

[[[0. 0. 1.]
  [0. 0. 1.]
  [0. 0. 1.]
  [1. 1. 1.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 1.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 1. 1.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 1.]
  [0. 0. 0.]]

 [[0. 0. 0.]
  [0. 0. 0.]
  [0. 0. 1.]
  [0. 0. 0.]]]


In [63]:
sess.close()

### 2

Solutions follow the seed 123.

#### 2.1
Use the Bernoulli likelihood function

$L\left(W,b\right)	=\prod_{i=1}^{n}f\left(y_{i};x,W,b\right)
	=\prod_{i=1}^{n}p\left(x_{i}\right)^{y_{i}}\left(1-p\left(x_{i}\right)\right)^{1-y_{i}}$
    
$\log l\left(W,b,x;y\right)$

$=\log\left[\prod_{i=1}^{n}p\left(x_{i}\right)^{y_{i}}\left(1-p\left(x_{i}\right)\right)^{1-y_{i}}\right]$

$=\sum_{i=1}^{n}\log p\left(x_{i}\right)^{y_{i}}+\sum_{i=1}^{n}\log\left(1-p\left(x_{i}\right)\right)^{1-y_{i}}$

$=\sum_{i=1}^{n}y_{i}\log\left[p\left(x_{i}\right)\right]+\sum_{i=1}^{n}\left(1-y_{i}\right)\log\left(1-p\left(x_{i}\right)\right)$

$=\sum_{i=1}^{n}\log\left(1-p\left(x_{i}\right)\right)+\sum_{i=1}^{n}y_{i}\left(\log\left[p\left(x_{i}\right)\right]-\log\left(1-p\left(x_{i}\right)\right)\right)$

$=\sum_{i=1}^{n}\log\left(1-p\left(x_{i}\right)\right)+\sum_{i=1}^{n}y_{i}\left(\frac{\log\left[p\left(x_{i}\right)\right]}{\log\left(1-p\left(x_{i}\right)\right)}\right)$

$=\sum_{i=1}^{n}\log\left(\frac{e^{-\left(W^{T}x_{i}+b\right)}}{1+e^{-\left(W^{T}x_{i}+b\right)}}\right)+\sum_{i=1}^{n}y_{i}\log\left(\frac{\frac{1}{1+e^{-\left(W^{T}X+b\right)}}}{\frac{e^{-\left(W^{T}x_{i}+b\right)}}{1+e^{-\left(W^{T}x_{i}+b\right)}}}\right)$

$=\sum_{i=1}^{n}\log\left(\frac{1}{1+e^{W^{T}x_{i}+b}}\right)+\sum_{i=1}^{n}y_{i}\left[\log\left(\frac{1}{e^{-\left(W^{T}x_{i}+b\right)}}\right)\right]$

$=-\sum_{i=1}^{n}\log\left(1+e^{W^{T}x_{i}+b}\right)+\sum_{i=1}^{n}y_{i}\left[\log\left(e^{W^{T}x_{i}+b}\right)\right]$

$=-\sum_{i=1}^{n}\log\left(1+e^{W^{T}x_{i}+b}\right)+\sum_{i=1}^{n}y_{i}\left(W^{T}x_{i}+b\right)$

In [107]:
import numpy as np
import tensorflow as tf

In [64]:
x = tf.placeholder(tf.float32, [None, 6])
W = tf.Variable(tf.zeros([6, 1]))
b = tf.Variable(tf.zeros([1]))
y = tf.nn.softmax(tf.matmul(x,W) + b)

The loss expression in tensorflow is

In [ ]:
loss = tf.reduce_sum(-tf.log(1 + tf.exp(tf.matmul(x,W) + b)) + y * (tf.matmul(x,W) + b), axis=[1])

#### 2.2

In [108]:
xtrain = np.load('logistic_xtrain.npy')
xtest = np.load('logistic_xtest.npy')
ytrain = np.load('logistic_ytrain.npy')
ytest = np.load('logistic_ytest.npy')
import random
random.seed(123)

In [109]:
x = tf.placeholder(tf.float32, [None, 6]) # n x 6
W = tf.Variable(tf.zeros([6,1])) # 6 x 1
b = tf.Variable(tf.zeros([1])) # broadcasted when adding to xW
y = tf.nn.softmax(tf.matmul(x,W) + b) # based on whatever is fed through x

ytrue = tf.placeholder(tf.float32, [None,1])
loss = tf.reduce_sum(-tf.log(1 + tf.exp(tf.matmul(x,W) + b)) +  y * (tf.matmul(x,W) + b), axis=[0])
train_step = tf.train.GradientDescentOptimizer(0.05).minimize(loss, var_list = (W, b))

In [110]:
# This function randomly samples from the data sets that are input as arguments
def next_batch(datx, daty, n):
    indx = np.random.randint(len(datx), size=n)
    return datx[indx], daty[indx]

In [111]:
sess = tf.InteractiveSession()
tf.global_variables_initializer().run()

for _ in range(5000):
    batch_xs, batch_ys = next_batch(xtrain, ytrain, 350)
    sess.run(train_step, feed_dict = {x: batch_xs, ytrue: batch_ys})

The estimates of W and b are as follows:

In [112]:
W.eval()

array([[ 376.32022 ],
       [ 304.4048  ],
       [  18.014465],
       [-360.78915 ],
       [1110.5889  ],
       [ 219.66277 ]], dtype=float32)

In [113]:
b.eval()

array([-87491.25], dtype=float32)

#### 2.3

In [114]:
sess.run(loss, feed_dict = {x: xtest, y: ytest}) # compute loss for test values

array([-17894284.], dtype=float32)

The loss is on the order of  $ 10^7$

#### 2.4

In [115]:
W_true = tf.constant([1,1,2,3,5,8], dtype = tf.float32)
b_true = tf.constant([-1], dtype = tf.float32)

In [116]:
tf.square(b - b_true).eval()

array([7.654544e+09], dtype=float32)

In [117]:
tf.square(tf.norm(W - W_true)).eval()

9812487.0

The sum of square error is on the order of $7\times 10^9$

In [119]:
(tf.square(b - b_true) + tf.square(tf.norm(W - W_true))).eval()

array([7.6643564e+09], dtype=float32)

In [120]:
sess.close()

#### 2.5

In [121]:
normal_xtrain = np.load('normal_xtrain.npy')
normal_xtest = np.load('normal_xtest.npy')
normal_ytrain = np.load('normal_ytrain.npy')
normal_ytest = np.load('normal_ytest.npy')

In [122]:
mn = tf.Variable(tf.zeros([3]))
var = tf.Variable([1,1,1], dtype = tf.float32)
x = tf.placeholder(tf.float32)
y = tf.placeholder(tf.float32)

dist = tf.distributions.Normal(loc = mn, scale = var)

cross_entropy = -tf.reduce_sum(y * tf.log(dist.prob(x)))
train_step2 = tf.train.AdagradOptimizer(0.01).minimize(cross_entropy)

In [123]:
sess2 = tf.InteractiveSession()
tf.global_variables_initializer().run()

for _ in range(5000):
    batch_xs, batch_ys = next_batch(normal_xtrain, normal_ytrain, 350)
    sess2.run(train_step2, feed_dict = {x: batch_xs, y: batch_ys})

In [124]:
mn.eval()

array([-0.95460665,  0.00838992,  1.0441424 ], dtype=float32)

The estimates for $\mu_1, \mu_2, \mu_3$ (respectively) with the training data are approximately -0.95, 0.008, and 1.044.

In [125]:
var.eval()

array([0.7144269, 1.000675 , 1.740305 ], dtype=float32)

The estimates for variances $\sigma_1^2, \sigma_2^2, \sigma_3^2$ (respectively) are approximately 0.71, 1.00, 1.74.

#### 2.6

The cross-entropy for the test data is about 777.

In [126]:
sess2.run(cross_entropy, feed_dict = {x: normal_xtest, y: normal_ytest})

777.18256

#### 2.7

In [127]:
mn_true = tf.constant([-1,0,3], dtype=tf.float32)
var_true = tf.constant([0.5,1,1.5], dtype=tf.float32)

The squared errors for each mean parameter and its estimate is listed below

In [128]:
tf.square(mn_true - mn).eval()

array([2.0605561e-03, 7.0390677e-05, 3.8253791e+00], dtype=float32)

The squared errors for each variance parameter and its estimate is listed below

In [129]:
tf.square(var_true - var).eval()

array([4.5978885e-02, 4.5557505e-07, 5.7746466e-02], dtype=float32)

The total sum of square error is listed below:

In [130]:
tf.reduce_sum(tf.square(mn_true - mn) + tf.square(var_true - var)).eval()

3.9312358

In [131]:
sess2.run(tf.one_hot(tf.argmax(dist.prob(x), axis=1), depth = 3),{x:normal_xtest})

array([[1., 0., 0.],
       [0., 0., 1.],
       [1., 0., 0.],
       ...,
       [1., 0., 0.],
       [0., 1., 0.],
       [1., 0., 0.]], dtype=float32)

In [132]:
probs = sess2.run(tf.equal(y, tf.one_hot(tf.argmax(dist.prob(x), axis=1), depth = 3)), 
                  {x: normal_xtest, y: normal_ytest})

In [133]:
1 - probs.mean()

0.18933333333333335

The classification error is about 18.9%

In [37]:
sess2.close()

### 3

I followed the Linear models tutorial. All functions and classes are from the tutorial code. The only addditional code I needed was to define FLAGS and \_SHUFFLE_BUFFER, as well as importing tempfile. Instead of nsteps, the batch_size and numepochs are changed

In [134]:
import urllib.request
import tensorflow as tf

In [135]:
def download_and_clean_file(filename, url):
    """Downloads data from url, and makes changes to match the CSV format."""
    temp_file, _ = urllib.request.urlretrieve(url)
    with tf.gfile.Open(temp_file, 'r') as temp_eval_file:
        with tf.gfile.Open(filename, 'w') as eval_file:
            for line in temp_eval_file:
                line = line.strip()
                line = line.replace(', ', ',')
                if not line or ',' not in line:
                    continue
                if line[-1] == '.':
                    line = line[:-1]
                line += '\n'
                eval_file.write(line)
    tf.gfile.Remove(temp_file)


In [136]:
download_and_clean_file('tut_train.csv','https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data')
download_and_clean_file('tut_test.csv','https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.test')

In [137]:
# Continuous columns
age = tf.feature_column.numeric_column('age')
education_num = tf.feature_column.numeric_column('education_num')
capital_gain = tf.feature_column.numeric_column('capital_gain')
capital_loss = tf.feature_column.numeric_column('capital_loss')
hours_per_week = tf.feature_column.numeric_column('hours_per_week')

education = tf.feature_column.categorical_column_with_vocabulary_list(
    'education', [
        'Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
        'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', 'Prof-school',
        '5th-6th', '10th', '1st-4th', 'Preschool', '12th'])

marital_status = tf.feature_column.categorical_column_with_vocabulary_list(
    'marital_status', [
        'Married-civ-spouse', 'Divorced', 'Married-spouse-absent',
        'Never-married', 'Separated', 'Married-AF-spouse', 'Widowed'])

relationship = tf.feature_column.categorical_column_with_vocabulary_list(
    'relationship', [
        'Husband', 'Not-in-family', 'Wife', 'Own-child', 'Unmarried',
        'Other-relative'])

workclass = tf.feature_column.categorical_column_with_vocabulary_list(
    'workclass', [
        'Self-emp-not-inc', 'Private', 'State-gov', 'Federal-gov',
        'Local-gov', '?', 'Self-emp-inc', 'Without-pay', 'Never-worked'])

# To show an example of hashing:
occupation = tf.feature_column.categorical_column_with_hash_bucket(
    'occupation', hash_bucket_size=1000)

# Transformations.
age_buckets = tf.feature_column.bucketized_column(
    age, boundaries=[18, 25, 30, 35, 40, 45, 50, 55, 60, 65])

education_x_occupation = tf.feature_column.crossed_column(
    ['education', 'occupation'], hash_bucket_size=1000)

age_buckets_x_education_x_occupation = tf.feature_column.crossed_column(
    [age_buckets, 'education', 'occupation'], hash_bucket_size=1000)

In [138]:
base_columns = [
    education, marital_status, relationship, workclass, occupation,
    age_buckets,
]
crossed_columns = [
    tf.feature_column.crossed_column(
        ['education', 'occupation'], hash_bucket_size=1000),
    tf.feature_column.crossed_column(
        [age_buckets, 'education', 'occupation'], hash_bucket_size=1000),
]

import tempfile

model_dir = tempfile.mkdtemp()
model = tf.estimator.LinearClassifier(
    model_dir=model_dir, feature_columns=base_columns + crossed_columns)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': '/var/folders/fq/10gmlvj140j067qh3_zsfhbh0000gn/T/tmpnu10a7el', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x10f9ce0f0>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


In [139]:
from random import shuffle

_CSV_COLUMNS = [
    'age', 'workclass', 'fnlwgt', 'education', 'education_num',
    'marital_status', 'occupation', 'relationship', 'race', 'gender',
    'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
    'income_bracket'
]

_CSV_COLUMN_DEFAULTS = [[0], [''], [0], [''], [0], [''], [''], [''], [''], [''],
                        [0], [0], [0], [''], ['']]

_SHUFFLE_BUFFER = 1500

def input_fn(data_file, num_epochs, shuffle, batch_size):
    """Generate an input function for the Estimator."""
    assert tf.gfile.Exists(data_file), (
      '%s not found. Please make sure you have either run data_download.py or '
      'set both arguments --train_data and --test_data.' % data_file)

    def parse_csv(value):
        print('Parsing', data_file)
        columns = tf.decode_csv(value, record_defaults=_CSV_COLUMN_DEFAULTS)
        features = dict(zip(_CSV_COLUMNS, columns))
        labels = features.pop('income_bracket')
        return features, tf.equal(labels, '>50K')

    # Extract lines from input files using the Dataset API.
    dataset = tf.data.TextLineDataset(data_file)

    if shuffle:
        dataset = dataset.shuffle(buffer_size=_SHUFFLE_BUFFER)

    dataset = dataset.map(parse_csv, num_parallel_calls=5)

    # We call repeat after shuffling, rather than before, to prevent separate
    # epochs from blending together.
    dataset = dataset.repeat(num_epochs)
    dataset = dataset.batch(batch_size)

    iterator = dataset.make_one_shot_iterator()
    features, labels = iterator.get_next()
    return features, labels

In [140]:
model.train(input_fn=lambda: input_fn('tut_train.csv', 5, True, 25))

Parsing tut_train.csv
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1 into /var/folders/fq/10gmlvj140j067qh3_zsfhbh0000gn/T/tmpnu10a7el/model.ckpt.
INFO:tensorflow:loss = 17.32868, step = 1
INFO:tensorflow:global_step/sec: 134.778
INFO:tensorflow:loss = 5.653425, step = 101 (0.744 sec)
INFO:tensorflow:global_step/sec: 287.286
INFO:tensorflow:loss = 12.125192, step = 201 (0.349 sec)
INFO:tensorflow:global_step/sec: 214.97
INFO:tensorflow:loss = 10.929335, step = 301 (0.465 sec)
INFO:tensorflow:global_step/sec: 215.769
INFO:tensorflow:loss = 5.7738132, step = 401 (0.464 sec)
INFO:tensorflow:global_step/sec: 288.636
INFO:tensorflow:loss = 7.6536865, step = 501 (0.346 sec)
INFO:tensorflow:global_step/sec: 275.261
INFO:tensorflow:loss = 6.5063953, step = 601 (0.364

In [141]:
results = model.evaluate(input_fn=lambda: input_fn('tut_test.csv', 5, False, 25))
for key in sorted(results):
    print('%s: %s' % (key, results[key]))

Parsing tut_test.csv
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2018-04-23-03:37:01
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /var/folders/fq/10gmlvj140j067qh3_zsfhbh0000gn/T/tmpnu10a7el/model.ckpt-6513
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-04-23-03:37:16
INFO:tensorflow:Saving dict for global step 6513: accuracy = 0.83575946, accuracy_baseline = 0.76377374, auc = 0.8832832, auc_precision_recall = 0.69409615, average_loss = 0.3521632, global_step = 6513, label/mean = 0.23622628, loss = 8.801918, prediction/mean = 0.23948163
accuracy: 0.83575946
accuracy_baseline: 0.76377374
auc: 0.8832832
auc_precision_recall: 0.69409615
average_loss: 0.3521632
global_step: 6513
label/mean: 0.23622628
loss: 8.801918
prediction/mean: 0.23948163


In [46]:
#This part of tutorial printed too many things and froze the notebook so I commented it out

# pred_iter = model.predict(input_fn=lambda: input_fn('tut_test.csv', 1, False, 1))
# for pred in pred_iter:
#     print(pred['classes'])

In [47]:
model = tf.estimator.LinearClassifier(
    model_dir=model_dir, feature_columns=base_columns + crossed_columns,
    optimizer=tf.train.FtrlOptimizer(
        learning_rate=0.1,
        l1_regularization_strength=1.0,
        l2_regularization_strength=1.0))

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': '/var/folders/fq/10gmlvj140j067qh3_zsfhbh0000gn/T/tmp91l7mz4_', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': None, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_service': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x1267e0d68>, '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


### 4

#### 4.1

See cloud account

#### 4.2

In [57]:
mn = tf.Variable(tf.zeros([3]))
var = tf.Variable([1,1,1], dtype = tf.float32)
x = tf.placeholder(tf.float32)
y = tf.placeholder(tf.float32)

dist = tf.distributions.Normal(loc = mn, scale = var)

cross_entropy = -tf.reduce_sum(y * tf.log(dist.prob(x)))
train_step3 = tf.train.AdagradOptimizer(0.01).minimize(cross_entropy)

sess3 = tf.InteractiveSession()
tf.global_variables_initializer().run()

for _ in range(5000):
    batch_xs, batch_ys = next_batch(normal_xtrain, normal_ytrain, 350)
    sess3.run(train_step3, feed_dict = {x: batch_xs, y: batch_ys})

pred = tf.one_hot(tf.argmax(dist.prob(x), axis=1), depth = 3)
    
tf.saved_model.simple_save(
    sess3,
    "./davidguo_normal_trained",
    inputs = {"x": x},
    outputs = {"pred": pred},
    legacy_init_op=None
)

INFO:tensorflow:Assets added to graph.
INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: b'./davidguo_normal_trained/saved_model.pb'


In [50]:
sess3.close()

In [51]:
print(tf.VERSION)

1.6.0


#### 4.3

I used the GUI to create a bucket

#### 4.4
I used the GUI to create a new model.

To create a new version with runtime 1.6, I typed

into the google cloud command line. I checked with 

which showed

#### 4.5

The json file is included in the submission.

#### 4.6

The model thinks $x=4$ is in the 3rd cluster. The following is the shell input and output: